examples/fungible-token/tests/contracts/defi/src/lib.rs

```diff
@@ -0,0 +1,53 @@
+/*!
+Some hypothetical DeFi contract that will do smart things with the transferred tokens
+*/
+use near_sdk::borsh::{self, BorshDeserialize, BorshSerialize};
+use near_sdk::json_types::{ValidAccountId, U128};
+use near_sdk::{env, ext_contract, log, near_bindgen, setup_alloc, PromiseOrValue};
+
+setup_alloc!();
+
+#[near_bindgen]
+#[derive(BorshDeserialize, BorshSerialize)]
+pub struct DeFi {}
+
+impl Default for DeFi {
+    fn default() -> Self {
+        Self {}
+    }
+}
+
+#[ext_contract(me)]
+pub trait CallMyself {
+    fn value_please(&self, amount_to_return: String) -> U128;
+}
+
+#[near_bindgen]
+#[allow(unused_variables)]
+impl DeFi {
+    /// If given `msg: "take-my-money", immediately returns U128::From(0)
+    /// Otherwise, makes a cross-contract call to own `value_please` function, passing `msg`
+    /// value_please will attempt to parse `msg` as an integer and return a U128 version of it
+    pub fn ft_on_transfer(
+        &self,
+        sender_id: ValidAccountId,
+        amount: U128,
+        msg: String,
+    ) -> PromiseOrValue<U128> {
+        log!("in ft_on_transfer, msg = {:?}", msg);
+        match msg.as_str() {
+            "take-my-money" => PromiseOrValue::Value(U128::from(0)),
+            _ => {
+                let prepaid_gas = env::prepaid_gas();
+                let account_id = env::current_account_id();
+                me::value_please(msg, &account_id, 0, prepaid_gas / 4).into()
```

Comment:
Should this use `prepaid_gas / 4`? It seems a bit arbitrary. I stole it from the High Level Cross Contracts example. I'm not sure what else to set it to, and this worked, so I left it.

Zooming out, in the general case, how can a contract developer determine a reasonable `gas` amount to attach to a cross contract call?

Reply:
@amgando and I discussed this the other day.  It would be great to have a `gas_left` from `env`, which you can do currently by subtracting prepaid and from used, but you need to make sure to leave a buffer for the current contract.

Reply:
The issue is that promises are not resolved until the end of the execution. And you also need to keep some for writing the serialized contract state back.